File: BinarySearchTreeIterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        self.stk = []
        while root:
            self.stk.append(root)
            root = root.left

    def hasNext(self):
        return self.stk
        

    def next(self):
        curr = self.stk.pop()
        val = curr.val
        curr = curr.right
        while curr:
            self.stk.append(curr)
            curr = curr.left
        return val
        

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())

File: binarySearchTreeIterator.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class BSTIterator(root: TreeNode?) {
    val stk = Stack<TreeNode>()
    var next: TreeNode? = null
    init{
        next = root
        while(next?.left != null){
            stk.push(next!!)
            next = next?.left
        }
    }
    fun next(): Int {
        val tmp = next!!
        if(next?.right != null){
            next = next?.right
            while(next?.left != null){
                stk.push(next!!)
                next = next?.left
            }
        }
        else if(stk.isNotEmpty())
            next = stk.pop()
        else
            next = null
        return tmp.`val`
    }

    fun hasNext(): Boolean {
        return next != null
    }

}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
